### PR TITLE
dnsdist: Implement LuaFFIRule, LuaFFIAction and LuaFFIResponseAction

### LINKED ISSUE #7617: dnsdist: Lua FFI interface to selectors and actions
- Program: dnsdist 
 - Issue type: Feature request 

### Short description

It would be nice to have fast `FFI` equivalents to `LuaAction` and `LuaResponseAction`, but also a way to write selector in `Lua`.

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

This PR implements:
- `LuaFFIRule`: to be able to write a query/response selector in `Lua` using the FFI interface ;
- `LuaFFIAction`: to be able to write actions in `Lua` using the FFI interface ;
- `LuaFFIResponseAction`: to be able to write response actions in `Lua` using the FFI interface ;
- `LuaRule()`: to be able to write a query/response selector in `Lua` (non-FFI version) ;
- `setServerPolicyLuaFFI`: to be able to write server load-balancing policies in `Lua` using the FFI interface.

It also refactor the code around the server selection policies and backends, slightly speeding up some of the built-in C++ policies in the process, and adding unit tests for them.

Closes #7617.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
I also spotted some remaining debug messages in the code :-)

**Comment by rgacogne** [comment]
Rebased to fix conflicts.

File: pdns/dnsdistdist/docs/reference/netmask.rst [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  .. method:: Netmask:getBits() -> int
+
+    Return the number of bits of this netmask, for example ``24`` for ``192.0.2.0/24``.
+
+  .. method:: Netmask:getMaskedNetwork() -> ComboAddress
+
+    Return a :class:`ComboAddress` object representing the base network of this netmask object after masking any additional bits if necessary (for example ``192.0.2.0`` if the netmask was constructed with ``newNetmask('192.0.2.1/24')).
+
+  .. method:: Netmask:isEmpty() -> bool
```
Comment (chbruyand) [inline pdns/dnsdistdist/docs/reference/netmask.rst]: I don't think this is the correct name (`empty()`) ?

File: pdns/dnsdist-lua-bindings.cc [hunk 2]
```diff
@@ -172,6 +174,29 @@ void setupLuaBindings(bool client)
 
   /* ComboAddress */
   g_lua.writeFunction("newCA", [](const std::string& name) { return ComboAddress(name); });
+  g_lua.writeFunction("newCAFromRaw", [](const std::string& raw, boost::optional<uint16_t> port) {
```
Comment (chbruyand) [inline pdns/dnsdist-lua-bindings.cc]: Maybe document the new functions introduced here ?

File: pdns/dnsdist.hh [hunk 3]
```diff
@@ -849,17 +852,19 @@ struct DownstreamState
...
   string getNameWithAddr() const {
-    if (name.empty()) {
-      return remote.toStringWithPort();
+    return nameWithAddr;
+  }
+  void setName(const std::string& newName)
+  {
+    name = newName;
+    if (newName.empty()) {
+      nameWithAddr = newName.empty() ? remote.toStringWithPort() : (name + " (" + remote.toStringWithPort()+ ")");
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsdist.hh]: Nice catch!

File: pdns/dnsdist-lua-bindings.cc (hunk 2 above)
Comment (rgacogne) [reply pdns/dnsdist-lua-bindings.cc]: This function is very specific to what we do with our own Lua FFI code, I'll ponder if it's worth documenting it.

**Comment by rgacogne** [comment]
Thanks a lot for the review!

**Pull request merged**